File: backend/src/api/routes/jobs.py
from fastapi import APIRouter, Depends, HTTPException
from fastapi.concurrency import run_in_threadpool
from typing import List
from sqlalchemy.ext.asyncio import AsyncSession
from src.api.db.session import get_db
from src.api.services.job_service import JobService
from src.api.schemas.job import (
    JobResponse, JobCreate, JobUpdate, JobImproveRequest, 
    JobDraftRequest, JobReviewRequest, JobExtendDeadline, BulkLifecycleRequest
)
from src.api.core.dependencies import get_current_user
from src.api.models.user import User

router = APIRouter()

@router.get("/ping")
async def ping(db: AsyncSession = Depends(get_db)):
    from sqlalchemy import text
    await db.execute(text("SELECT 1"))
    return {"message": "pong"}

@router.get("/stats/count")
async def get_jobs_stats(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Returns the total number of jobs created"""
    job_service = JobService(db)
    count = await job_service.get_total_jobs_count()
    return {"total_jobs": count}

@router.get("/stats/dashboard")
async def get_dashboard_stats(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Returns dashboard statistics including pending actions"""
    job_service = JobService(db)
    return await job_service.get_dashboard_stats(user_id=current_user.id)

from src.api.models.job import JobStatus

@router.get("/public", response_model=List[JobResponse])
async def read_public_jobs(
    skip: int = 0,
    limit: int = 100,
    db: AsyncSession = Depends(get_db)
):
    """Public endpoint for fetching published jobs (no authentication required)"""
    job_service = JobService(db)
    return await job_service.get_jobs(skip=skip, limit=limit, status=JobStatus.PUBLISHED.value)
@router.get("", response_model=List[JobResponse])
async def read_jobs(
    skip: int = 0,
    limit: int = 100,
    status: str = None,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Authenticated endpoint for fetching jobs created by the current user"""
    job_service = JobService(db)
    # Filter by current user to ensure they only see their own generated jobs
    return await job_service.get_my_jobs(user_id=current_user.id, skip=skip, limit=limit, status=status)

@router.post("", response_model=JobResponse)
async def create_job(
    job_data: JobCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Create a new job posting"""
    job_service = JobService(db)
    return await job_service.create_job(job_in=job_data, user_id=current_user.id)

@router.post("/generate-draft")
async def generate_draft(
    draft_data: JobDraftRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Generate a structured job description draft using AI"""
    job_service = JobService(db)
    return await job_service.generate_draft(draft_data)

@router.get("/{job_id}", response_model=JobResponse)
async def read_job(
    job_id: int,
    db: AsyncSession = Depends(get_db)
):
    """
    Get a specific job by ID.
    Publicly accessible to allow candidates to view job details.
    """
    job_service = JobService(db)
    job = await job_service.get_job(job_id)
    if not job:
        from fastapi import HTTPException
        raise HTTPException(status_code=404, detail="Job not found")
    return job


@router.put("/{job_id}", response_model=JobResponse)
async def update_job(
    job_id: int,
    job_data: JobUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Update a specific job post manually"""
    job_service = JobService(db)
    job = await job_service.update_job(job_id, job_data)
    if not job:
        from fastapi import HTTPException
        raise HTTPException(status_code=404, detail="Job not found")
    return job


@router.post("/{job_id}/improve", response_model=JobResponse)
async def improve_job(
    job_id: int,
    request: JobImproveRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Refine job post using AI based on HR feedback"""
    job_service = JobService(db)
    job = await job_service.improve_job(job_id, request.feedback)
    if not job:
        from fastapi import HTTPException
        raise HTTPException(status_code=404, detail="Job not found")
    return job


@router.post("/{job_id}/publish", response_model=JobResponse)
async def publish_job(
    job_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    job_service = JobService(db)
    job = await job_service.publish_job(job_id, current_user.id)
    if not job:
        from fastapi import HTTPException
        raise HTTPException(status_code=404, detail="Job not found")
    return job

from src.api.services.email_service import EmailService
@router.post("/{job_id}/send-to-manager")
async def send_to_manager(
    job_id: int,
    role: str = None,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Send job details to a specific team lead or Operation Manager."""
    job_service = JobService(db)
    job = await job_service.get_job(job_id)
    if not job:
        raise HTTPException(status_code=404, detail="Job not found")

    details = (
        f"Title: {job.title}\n"
        f"Location: {job.location}\n"
        f"Type: {job.job_type}\n"
        f"Experience: {job.experience_level}\n"
        f"Department: {job.department}\n"
        f"\nDescription:\n{job.description}\n"
    )

    # Map role to email
    from src.api.core.config import settings
    recipient_email = settings.OPERATIONS_MANAGER_EMAIL
    
    if role == "ai":
        recipient_email = settings.AI_LEAD_EMAIL
    elif role == "uiux":
        recipient_email = settings.UIUX_LEAD_EMAIL
    elif role == "seo":
        recipient_email = settings.SEO_LEAD_EMAIL
    elif role == "webdev":
        recipient_email = settings.WEB_DEV_LEAD_EMAIL

    try:
        review_url = f"{settings.FRONTEND_URL}/review-job/{job_id}"
        await EmailService.send_job_to_manager(job.title, details, review_url=review_url, recipient_email=recipient_email)
        return {"message": f"Email sent successfully to {role or 'manager'}"}
    except Exception as e:
        import traceback
        traceback.print_exc()
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/{job_id}/extend-deadline", response_model=JobResponse)
async def extend_deadline(
    job_id: int,
    request: JobExtendDeadline,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Extend the job application deadline"""
    job_service = JobService(db)
    job = await job_service.extend_deadline(job_id, request.new_deadline, current_user.id)
    if not job:
        raise HTTPException(status_code=404, detail="Job not found")
    return job

@router.post("/{job_id}/close", response_model=JobResponse)
async def close_job(
    job_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Manually close a job post"""
    job_service = JobService(db)
    job = await job_service.close_job(job_id)
    if not job:
        raise HTTPException(status_code=404, detail="Job not found")
    return job

@router.post("/{job_id}/reopen", response_model=JobResponse)
async def reopen_job(
    job_id: int,
    request: JobExtendDeadline,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Reopen a closed or expired job with a new deadline"""
    job_service = JobService(db)
    job = await job_service.reopen_job(job_id, request.new_deadline, current_user.id)
    if not job:
        raise HTTPException(status_code=404, detail="Job not found")
    return job

@router.post("/{job_id}/archive", response_model=JobResponse)
async def archive_job(
    job_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Archive a job post (final state)"""
    job_service = JobService(db)
    job = await job_service.archive_job(job_id)
    if not job:
        raise HTTPException(status_code=404, detail="Job not found")
    return job

@router.post("/bulk/lifecycle")
async def bulk_lifecycle_action(
    request: BulkLifecycleRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Bulk action for job lifecycle (close, archive, extend)"""
    job_service = JobService(db)
    results = await job_service.bulk_action(
        job_ids=request.job_ids,
        action=request.action,
        new_deadline=request.new_deadline,
        user_id=current_user.id
    )
    return {"message": "Bulk action completed", "processed": len(results)}

File: frontend/src/lib/api/jobs.ts
import { apiClient } from './client';
import type {
    Job,
    JobStatus,
    JobIntent,
    AIJobDraft,
    PaginatedResponse,
    ApiResponse,
} from '@/lib/types';

/**
 * Job API endpoints
 */

/**
 * Helper to map backend job response to frontend Job type
 */
const mapJob = (job: any): Job => ({
    id: job.id.toString(),
    title: job.title,
    description: job.description,
    short_description: job.short_description,
    department: job.department,
    location: job.location,
    location_type: job.location_type,
    company_name: job.company_name,
    job_type: job.job_type,
    type: job.job_type, // Compatibility field
    experience_level: job.experience_level,
    salary_min: job.salary_min,
    salary_max: job.salary_max,
    salary_currency: job.salary_currency,
    salary_period: job.salary_period,
    salary_range: job.salary_range,
    required_skills: job.required_skills || [],
    preferred_skills: job.preferred_skills || [],
    requirements: job.requirements || [],
    preferred_qualifications: job.preferred_qualifications || [],
    benefits: job.benefits || [],
    application_url: job.application_url,
    status: job.status,
    effective_status: job.effective_status,
    // Backward compatibility fields
    desiredSkills: job.preferred_skills || [],
    candidateCount: job.application_count || 0,
    application_count: job.application_count || 0,
    pendingActionCount: 0,
    createdBy: job.created_by?.toString() || '',
    created_by: job.created_by,
    createdAt: job.created_at,
    created_at: job.created_at,
    publishedAt: job.published_at,
    published_at: job.published_at,
    closedAt: job.expires_at,
    expires_at: job.expires_at,
    manager_feedback: job.manager_feedback,
    closed_at: job.closed_at,
    archived_at: job.archived_at,
    reopened_at: job.reopened_at,
    extended_at: job.extended_at,
    extended_by: job.extended_by,
});

export const jobsApi = {
    /**
     * Get all jobs with optional filtering (requires authentication)
     */
    getAll: async (params?: {
        status?: string;
        department?: string;
        skip?: number;
        limit?: number;
    }): Promise<Job[]> => {
        const jobs = await apiClient.get<any[]>('/jobs', { params });
        return jobs.map(mapJob);
    },

    /**
     * Get published jobs (public, no authentication required)
     */
    getPublic: async (params?: {
        skip?: number;
        limit?: number;
    }): Promise<Job[]> => {
        const jobs = await apiClient.get<any[]>('/jobs/public', { params });
        return jobs.map(mapJob);
    },

    /**
     * Get single job by ID
     */
    getById: async (id: string): Promise<Job> => {
        const job = await apiClient.get<any>(`/jobs/${id}`);
        return mapJob(job);
    },

    /**
     * Create new job from intent
     */
    create: async (intent: JobIntent): Promise<ApiResponse<Job>> => {
        return apiClient.post<ApiResponse<Job>>('/jobs', intent);
    },

    /**
     * Generate an initial AI job draft without creating a record.
     * If `prompt` is provided, it is used as the primary AI instruction.
     */
    generateDraft: async (data: {
        title?: string;
        department?: string;
        location?: string;
        experience_level?: string;
        job_type?: string;
        prompt?: string;
    }): Promise<any> => {
        return apiClient.post<any>('/jobs/generate-draft', data);
    },

    /**
     * Trigger AI generation for job description
     */
    generateDescription: async (jobId: string): Promise<ApiResponse<AIJobDraft>> => {
        return apiClient.post<ApiResponse<AIJobDraft>>(`/jobs/${jobId}/generate`);
    },

    /**
     * Approve AI-generated job description
     */
    approveDraft: async (
        jobId: string,
        editedDescription?: string
    ): Promise<ApiResponse<Job>> => {
        return apiClient.post<ApiResponse<Job>>(`/jobs/${jobId}/approve`, {
            editedDescription,
        });
    },

    /**
     * Publish job to portal
     */
    publish: async (jobId: string): Promise<ApiResponse<Job>> => {
        return apiClient.post<ApiResponse<Job>>(`/jobs/${jobId}/publish`);
    },

    /**
     * Improve job description using AI based on feedback
     */
    improve: async (jobId: string, feedback: string): Promise<Job> => {
        return apiClient.post<Job>(`/jobs/${jobId}/improve`, { feedback });
    },

    /**
     * Update job details
     */
    update: async (jobId: string, updates: Partial<Job>): Promise<ApiResponse<Job>> => {
        return apiClient.put<ApiResponse<Job>>(`/jobs/${jobId}`, updates);
    },

    /**
     * Delete job
     */
    delete: async (jobId: string): Promise<ApiResponse<void>> => {
        return apiClient.delete<ApiResponse<void>>(`/jobs/${jobId}`);
    },

    /**
     * Close job posting
     */
    close: async (jobId: string): Promise<ApiResponse<Job>> => {
        return apiClient.post<ApiResponse<Job>>(`/jobs/${jobId}/close`);
    },

    /**
     * Extend job deadline
     */
    extendDeadline: async (jobId: string, newDeadline: string): Promise<ApiResponse<Job>> => {
        return apiClient.post<ApiResponse<Job>>(`/jobs/${jobId}/extend-deadline`, {
            new_deadline: newDeadline
        });
    },

    /**
     * Reopen a closed job
     */
    reopen: async (jobId: string, newDeadline: string): Promise<ApiResponse<Job>> => {
        return apiClient.post<ApiResponse<Job>>(`/jobs/${jobId}/reopen`, {
            new_deadline: newDeadline
        });
    },

    /**
     * Archive job post
     */
    archive: async (jobId: string): Promise<ApiResponse<Job>> => {
        return apiClient.post<ApiResponse<Job>>(`/jobs/${jobId}/archive`);
    },

    /**
     * Bulk lifecycle action
     */
    bulkLifecycle: async (data: {
        job_ids: string[];
        action: 'close' | 'archive' | 'extend';
        new_deadline?: string;
    }): Promise<{ message: string; processed: number }> => {
        return apiClient.post('/jobs/bulk/lifecycle', data);
    },

    /**
     * Send job details to Operation Manager or specific Team Lead
     */
    sendToManager: async (jobId: string, role?: string): Promise<{ message: string }> => {
        return apiClient.post<{ message: string }>(`/jobs/${jobId}/send-to-manager`, null, {
            params: { role }
        });
    },

    /**
     * Submit Operation Manager review
     */
    review: async (
        jobId: string,
        data: { status: JobStatus; feedback?: string }
    ): Promise<ApiResponse<Job>> => {
        return apiClient.post<ApiResponse<Job>>(`/jobs/${jobId}/review`, data);
    },

    /**
     * Get dashboard statistics
     */
    getStats: async (): Promise<{ total_jobs: number; pending_actions: number }> => {
        return apiClient.get<{ total_jobs: number; pending_actions: number }>('/jobs/stats/dashboard');
    },
};

